### PR TITLE
[1.10] File versioning

### PR DESCRIPTION
**Pull request by yurabakhtin**
Issue: https://github.com/humhub/cfiles/issues/121

### CODE REVIEW

File: controllers/VersionController.php [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+        return $this->asJson([
+            'html' => $versionsView->renderVersions(),
+            'isLast' => $versionsView->isLastPage(),
+        ]);
+    }
+
+    private function getFile(): File
+    {
+        $file = File::findOne(['id' => Yii::$app->request->get('id')]);
```
Comment (luke-) [inline controllers/VersionController.php:85]: Check `canView()` or `readable()` ?

File: module.json [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
         "docs",
-        "structre",
+        "structure",
         "organisation",
         "sharing"
     ],
-    "version": "0.13.2",
+    "version": "0.13.3",
     "humhub": {
         "minVersion": "1.8"
```
Comment (luke-) [inline module.json:14]: Min Version, 1.10

**Review by luke- — CHANGES_REQUESTED** [review]
@yurabakhtin Looks fine! Please check my comments.

Regarding permissions:

 Actually, anyone with read permission should be able to see older versions of the file. And anyone with a write permission should be able to reset a version.

A "ManageFiles" permission would probably not be necessary here? What do you think?

**Comment by yurabakhtin** [comment]
@luke- Commit https://github.com/humhub/cfiles/pull/138/commits/8fcc3b1c11b3e49e0ab5c1e2a98fc5d5de73b8f1.

> Actually, anyone with read permission should be able to see older versions of the file. And anyone with a write permission should be able to reset a version.
> 
> A "ManageFiles" permission would probably not be necessary here? What do you think?

I think the "Versioning" part is related to admin/back-office part. I mean these data should be allowed for users who can manage files. I think users with only read permissions should see only the latest/current version, because please imagine when some image or file was posted by mistake with some private date like password or it was some forbidden content that was replaced with correct content on next uploading, so read-only user should not see such fixed content.

File: module.json (hunk 2 above)
Comment (yurabakhtin) [reply module.json]: @luke- It was changed here - https://github.com/humhub/cfiles/pull/138/commits/8fcc3b1c11b3e49e0ab5c1e2a98fc5d5de73b8f1#diff-e6fdaf48ec339e2166f48043597d67d91464d12ef34fb958b12d6c93ba5f6fe3R14.

File: controllers/VersionController.php (hunk 1 above)
Comment (yurabakhtin) [reply controllers/VersionController.php]: @luke- It was done here - https://github.com/humhub/cfiles/pull/138/commits/8fcc3b1c11b3e49e0ab5c1e2a98fc5d5de73b8f1#diff-f5095e3e90393c3d9e281412a06219ceb36c8c01c0e918ad240225bcd9b0a8acR85-R98.

**Comment by luke-** [comment]
@yurabakhtin Good point. What do you think about `ManageFiles` users being able to delete versions as well?
I agree with that `canView()` users should be only allowed to access the latest version.
But `canEdit()` users should also see older versions, right?

**Comment by yurabakhtin** [comment]
@luke- Yes, I agree if user has permission like `ManageFiles` then it should be allowed to manage all old versions of all files, because the description of this permission is "Allows the user to modify or delete any files.", so no need to create any other new separate permission for works with versions.

> But `canEdit()` users should also see older versions, right?

I see the `canEdit()` of the CFiles module is `$cfile->content->canEdit()` so it is the same `ManageFiles`, i.e. it can be used for versions part as well.

**Comment by luke-** [comment]
@yurabakhtin There should be a difference `$cfile->content->canEdit() ` contains `ManageFiles` **and** `the file author`

**Comment by yurabakhtin** [comment]
@luke- 

> There should be a difference `$cfile->content->canEdit()` contains `ManageFiles` and `the file author`.

Yes, you are right the `$cfile->content->canEdit()` can return true in additional if:

- current user is author of the Content
- current user is admin and it is allowed to edit all content by admins(`humhub\modules\content\Module->adminCanEditAllContent = true;`)
- content's model has the method `canEdit` and it returns `true`;

**Comment by luke-** [comment]
@yurabakhtin Ok, so we have following new rights?

|                     | View Latest Version | View older versions | Delete older versions |
|---------------------|---------------------|---------------------|-----------------------|
| canView             | yes                 | no                  | no                    |
| canEdit             | yes                 | yes                 | no                    |
| ManageFiles / Admin |  yes                   | yes                     | yes                   |

**Review by luke- — APPROVED** [review]

**Review by luke- — APPROVED** [review]

**Pull request merged**